In [1]:
#Import packages
import numpy as np
import pandas as pd
import re

In [2]:
#Read the text files
with open('aljazeera-khashoggi.txt') as f:
    aljazeera=f.read()
with open('bbc-khashoggi.txt') as f:    
    bbc=f.read()
with open('breitbart-khashoggi.txt') as f:  
    breitbart=f.read() 
with open('cnn-khashoggi.txt') as f:   
    cnn=f.read()
with open('fox-khashoggi.txt') as f:   
    fox=f.read()

In [3]:
#Check one file
aljazeera

'Turkey, Istanbul - Turkish President Recep Tayyip Erdogan has said the murder of journalist Jamal Khashoggi at the kingdom\'s consulate in Istanbul was planned by Saudi officials days in advance. Addressing legislators from his Justice and Development Party (AK Party) on Tuesday, Erdogan detailed Khashoggi\'s disappearance and murder but stopped short of accusing Saudi royals of the "savage" killing that has caused global outrage. "On September 28, Khashoggi arrived at the Saudi Arabian consulate for him to sort out his wedding paperwork," Erdogan said during the speech in the Turkish parliament in the capital, Ankara. "It seems that at that time they [Saudi Arabian officials] started to plan a roadmap for his murder." He added that some Saudi officials left Turkey and travelled to Saudi Arabia, "indicating they planned the murder". Khashoggi, 59, a Washington Post columnist and critic of the powerful Saudi Crown Prince Mohammed bin Salman, disappeared after entering the Saudi consula

In [4]:
#Read the stopwords and transfer it into list form
stopword=pd.read_csv('stop_words.csv')
stopword = list(stopword["word"])

## Analysis on correlation by removing stopwords

In [5]:
#Turn the text into numerical data.
#Break up the sentences into individual words, scrub capitalization punctuation and numbers
def tokenize(text=None):
    '''
    Turn the text into numerical data.
    
    Parameters
    ------
    text: a string of txt file's name.
    
    Returns
    ------
    text_list: a list that contains each word of the txt file.
    '''
    text = text.lower()
    text = re.sub(r'[^\w\s]','',text)#Remove the punctuations
    text = re.sub(r'[0-9]+', '', text) #Remove the numbers
    text_list = text.split()#Split the word
    text_list = [word for word in text_list if word not in stopword]#Remove word in stopwords
    return text_list

#Test file 'aljazeera'
tokenize(aljazeera)

['turkey',
 'istanbul',
 'turkish',
 'president',
 'recep',
 'tayyip',
 'erdogan',
 'murder',
 'journalist',
 'jamal',
 'khashoggi',
 'kingdoms',
 'consulate',
 'istanbul',
 'planned',
 'saudi',
 'officials',
 'days',
 'advance',
 'addressing',
 'legislators',
 'justice',
 'development',
 'party',
 'ak',
 'party',
 'tuesday',
 'erdogan',
 'detailed',
 'khashoggis',
 'disappearance',
 'murder',
 'stopped',
 'short',
 'accusing',
 'saudi',
 'royals',
 'savage',
 'killing',
 'caused',
 'global',
 'outrage',
 'september',
 'khashoggi',
 'arrived',
 'saudi',
 'arabian',
 'consulate',
 'sort',
 'wedding',
 'paperwork',
 'erdogan',
 'speech',
 'turkish',
 'parliament',
 'capital',
 'ankara',
 'time',
 'saudi',
 'arabian',
 'officials',
 'started',
 'plan',
 'roadmap',
 'murder',
 'added',
 'saudi',
 'officials',
 'left',
 'turkey',
 'travelled',
 'saudi',
 'arabia',
 'indicating',
 'planned',
 'murder',
 'khashoggi',
 'washington',
 'post',
 'columnist',
 'critic',
 'powerful',
 'saudi',
 'cr

In [6]:
#Counts the number of words in the text
def wordsnumber(txt=None):
    '''
    Counts the number of words in the text
    
    Parameters
    ------
    txt: a string of txt file's name.
    
    Returns
    ------
    wordsnumber: a list that contains the number of each word in the txt file.
    '''
    #Create a dictionary
    d = dict()
    #Caculate the number of each word
    for word in tokenize(txt):
        if word in d:
            d[word][0] += 1
        else:
            d[word] = [1]
    return pd.DataFrame(d)
#Test file 'aljazeera'
wordsnumber(aljazeera)

turkey  istanbul  turkish  president  recep  tayyip  erdogan  murder  \
0       4         5        7          4      1       1       12       6   

   journalist  jamal  ...  aljubeir  terrible  tragedy  taha  ozhan  research  \
0           1      1  ...         1         1        1     1      1         1   

   director  institute  saudis  cooperation  
0         1          1       2            1  

[1 rows x 185 columns]

In [7]:
def gen_DTM(texts=None):
    '''
    Generate a document term matrix
    
    Parameters
    ------
    texts: a string of txt file's name.
    
    Returns
    ------
    DTM: a dataframe that contains the number of each word for all txt file.
    '''
    DTM = pd.DataFrame()
    for text in texts:
        entry = wordsnumber(text)
        DTM = DTM.append(pd.DataFrame(entry),ignore_index=True,sort=True) # Bind rows
    
    DTM.fillna(0, inplace=True) # Fill in  missing values with 0s 
    return DTM

# Test for all text files      
gen_DTM([aljazeera, bbc, breitbart, cnn, fox])


abdulaziz  absent  accident  accidentally  accidentallyerdogan  account  \
0        0.0     0.0       0.0           0.0                  0.0      0.0   
1        0.0     1.0       0.0           0.0                  0.0      1.0   
2        2.0     0.0       0.0           1.0                  0.0      1.0   
3        0.0     0.0       1.0           0.0                  1.0      0.0   
4        0.0     0.0       0.0           0.0                  0.0      0.0   

   accounts  accusation  accusing  acknowledged  ...  white  widely  withheld  \
0       0.0         0.0       1.0           0.0  ...    0.0     0.0       0.0   
1       2.0         0.0       0.0           1.0  ...    0.0     1.0       0.0   
2       0.0         1.0       1.0           0.0  ...    1.0     0.0       1.0   
3       0.0         0.0       0.0           0.0  ...    0.0     0.0       0.0   
4       0.0         0.0       0.0           0.0  ...    0.0     0.0       0.0   

   woods  world  worse  writer  yalova  yearold  yelova  
0    0.0    0.0    0.0     0.0     0.0      0.0     0.0  
1    0.0    1.0    0.0     0.0     0.0      1.0     0.0  
2    0.0    1.0    0.0     0.0     0.0      0.0     0.0  
3    0.0    0.0    0.0     0.0     1.0      0.0     0.0  
4    1.0    0.0    1.0     2.0     0.0      0.0     1.0  

[5 rows x 644 columns]

In [8]:
#Calculate the angle between two vectors
def cosine(a,b):
    '''
    Generate a function to calculate the angle between two vectors
    
    Parameters
    ------
    a,b: float 
         two vectors
    
    Returns
    ------
    cos: float
         the cosine value of the angle between the two vectors
    '''
    cos = np.dot(a,b)/(np.sqrt(np.dot(a,a)) * np.sqrt(np.dot(b,b))  )
    return cos



In [9]:
#Index the pandas dataframe to draw out a vector
D = gen_DTM([aljazeera, bbc, breitbart, cnn, fox])
a = D.iloc[0].values
b = D.iloc[1].values
c = D.iloc[2].values
d = D.iloc[3].values
e = D.iloc[4].values

In [10]:
#Create a new dictionary to assign the corresponding name for each vector 
df_nostop = {
'Aljazeera':a,
'BBC':b,
'Breitbart':c,
'CNN':d,
'Fox':e
} 
#Transfer the dictionary into dataframe form
df_nostop = pd.DataFrame(df_nostop)
#Make cosine correlation dataframe for the vectors
df_nostop_corr = df_nostop.corr(method=cosine).round(4)
#Dataframe for the correlation betetween vectors(not includes stopwords)
df_nostop_corr

Aljazeera     BBC  Breitbart     CNN     Fox
Aljazeera     1.0000  0.6792     0.6019  0.5330  0.6981
BBC           0.6792  1.0000     0.5976  0.5027  0.6473
Breitbart     0.6019  0.5976     1.0000  0.3784  0.5547
CNN           0.5330  0.5027     0.3784  1.0000  0.5357
Fox           0.6981  0.6473     0.5547  0.5357  1.0000

## Analysis on correlation including stopwords

In [11]:
#Turn the text into numerical data.
#Break up the sentences into individual words, scrub capitalization punctuation and numbers
def tokenize2(text=None):
    '''
    Turn the text into numerical data.
    
    Parameters
    ------
    text: a string of txt file's name.
    
    Returns
    ------
    text_list: a list that contains each word of the txt file.
    '''
    text = text.lower()
    text = re.sub(r'[^\w\s]','',text)#Remove the punctuations
    text = re.sub(r'[0-9]+', '', text) #Remove the numbers
    text_list2 = text.split()#Split the word
    return text_list2

#Test file 'aljazeera'
tokenize2(aljazeera)

['turkey',
 'istanbul',
 'turkish',
 'president',
 'recep',
 'tayyip',
 'erdogan',
 'has',
 'said',
 'the',
 'murder',
 'of',
 'journalist',
 'jamal',
 'khashoggi',
 'at',
 'the',
 'kingdoms',
 'consulate',
 'in',
 'istanbul',
 'was',
 'planned',
 'by',
 'saudi',
 'officials',
 'days',
 'in',
 'advance',
 'addressing',
 'legislators',
 'from',
 'his',
 'justice',
 'and',
 'development',
 'party',
 'ak',
 'party',
 'on',
 'tuesday',
 'erdogan',
 'detailed',
 'khashoggis',
 'disappearance',
 'and',
 'murder',
 'but',
 'stopped',
 'short',
 'of',
 'accusing',
 'saudi',
 'royals',
 'of',
 'the',
 'savage',
 'killing',
 'that',
 'has',
 'caused',
 'global',
 'outrage',
 'on',
 'september',
 'khashoggi',
 'arrived',
 'at',
 'the',
 'saudi',
 'arabian',
 'consulate',
 'for',
 'him',
 'to',
 'sort',
 'out',
 'his',
 'wedding',
 'paperwork',
 'erdogan',
 'said',
 'during',
 'the',
 'speech',
 'in',
 'the',
 'turkish',
 'parliament',
 'in',
 'the',
 'capital',
 'ankara',
 'it',
 'seems',
 'that'

In [12]:
#Counts the number of words in the text
def wordsnumber2(txt=None):
    '''
    Counts the number of words in the text
    
    Parameters
    ------
    txt: a string of txt file's name.
    
    Returns
    ------
    wordsnumber2: a list that contains the number of each word in the txt file.
    '''
    #Create a dictionary
    d = dict()
    #Caculate the number of words
    for word in tokenize2(txt):
        if word in d:
            d[word][0] += 1
        else:
            d[word] = [1]
    return pd.DataFrame(d)

wordsnumber2(aljazeera)

turkey  istanbul  turkish  president  recep  tayyip  erdogan  has  said  \
0       4         5        7          4      1       1       12    6     6   

   the  ...  saudis  know  very  well  knows  doing  namely  asking  full  \
0   38  ...       2     1     1     1      1      1       1       1     1   

   cooperation  
0            1  

[1 rows x 286 columns]

In [13]:
def gen_DTM2(texts=None):
    '''
    Generate a document term matrix
    
    Parameters
    ------
    texts: a string of txt file's name.
    
    Returns
    ------
    DTM2: a dataframe that contains the number of each word for all txt file.
    '''
    DTM2 = pd.DataFrame()
    for text in texts:
        entry = wordsnumber2(text)
        DTM2 = DTM2.append(pd.DataFrame(entry),ignore_index=True,sort=True) # Row bind
    
    DTM2.fillna(0, inplace=True) # Fill in any missing values with 0s (i.e. when a word is in one text but not another)
    return DTM2

# Test for all text files      
gen_DTM2([aljazeera, bbc, breitbart, cnn, fox])



a  abdulaziz  able  about  absent  accident  accidentally  \
0  11        0.0   0.0      2     0.0       0.0           0.0   
1  23        0.0   0.0      2     1.0       0.0           0.0   
2  11        2.0   0.0      2     0.0       0.0           1.0   
3  14        0.0   1.0      1     0.0       1.0           0.0   
4  14        0.0   0.0      4     0.0       0.0           0.0   

   accidentallyerdogan  according  account  ...  working  world  worse  would  \
0                  0.0        0.0      0.0  ...      1.0    0.0    0.0    1.0   
1                  0.0        1.0      1.0  ...      0.0    1.0    0.0    4.0   
2                  0.0        1.0      1.0  ...      0.0    1.0    0.0    1.0   
3                  1.0        0.0      0.0  ...      0.0    0.0    0.0    0.0   
4                  0.0        1.0      0.0  ...      0.0    0.0    1.0    2.0   

   writer  yalova  yearold  yelova  yet  your  
0     0.0     0.0      0.0     0.0  0.0   0.0  
1     0.0     0.0      1.0     0.0  0.0   3.0  
2     0.0     0.0      0.0     0.0  1.0   0.0  
3     0.0     1.0      0.0     0.0  1.0   0.0  
4     2.0     0.0      0.0     1.0  0.0   0.0  

[5 rows x 870 columns]

In [14]:
#Index the pandas dataframe to draw out a vector
D2 = gen_DTM2([aljazeera, bbc, breitbart, cnn, fox])
a2 = D2.iloc[0].values
b2 = D2.iloc[1].values
c2 = D2.iloc[2].values
d2 = D2.iloc[3].values
e2 = D2.iloc[4].values

In [15]:
#Create a new dictionary to assign the corresponding name for each vector 
df_stop = {
'Aljazeera':a2,
'BBC':b2,
'Breitbart':c2,
'CNN':d2,
'Fox':e2
} 
#Transfer the dictionary into dataframe form
df_stop = pd.DataFrame(df_stop)
#Make cosine correlation dataframe for the vectors
df_stop_corr = df_stop.corr(method=cosine).round(4)
#Dataframe for the correlation betetween vectors(includes stopwords in the text)
df_stop_corr

Aljazeera     BBC  Breitbart     CNN     Fox
Aljazeera     1.0000  0.8710     0.8401  0.7349  0.8458
BBC           0.8710  1.0000     0.8990  0.7440  0.8898
Breitbart     0.8401  0.8990     1.0000  0.6815  0.8687
CNN           0.7349  0.7440     0.6815  1.0000  0.7405
Fox           0.8458  0.8898     0.8687  0.7405  1.0000

In [16]:
#Dataframe for the correlation betetween vectors(not includes stopwords in the text)
df_nostop_corr

Aljazeera     BBC  Breitbart     CNN     Fox
Aljazeera     1.0000  0.6792     0.6019  0.5330  0.6981
BBC           0.6792  1.0000     0.5976  0.5027  0.6473
Breitbart     0.6019  0.5976     1.0000  0.3784  0.5547
CNN           0.5330  0.5027     0.3784  1.0000  0.5357
Fox           0.6981  0.6473     0.5547  0.5357  1.0000

# Conclusion
From the two correlation(similarity) tables, compared table with stopwords removed, the correlation between these reports is higher when including stopwords.
Case1(including stopwords in text):The report text of BBC and Breitbart, of BBC and Fox are extremly high(cosine value at 0.899 and 0.89 respectively). CNN's content and Breitbart's content has much difference, the cosine value is just 0.682.
Case2(removing stopwords in text): Fox and Aljazeera has highest similarities with cosine value at 0.698. However, the report text of BBC and Fox is not much correlated at this time. Also, CNN and Breitbart has the least similarity with cosine value at 0.378.
We can see that when we include stopwords in analysis of correlation, some information we get might be misleading.
We may get a high correlation result because of the high frequency use of stopwords, which are common to the English Language but make little sense of the main purpose of the text.